src/search/hnsw_indexer.h

```diff
@@ -0,0 +1,395 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <cmath>
+#include <memory>
+#include <random>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include <queue>
+
+#include "db_util.h"
+#include "parse_util.h"
+#include "search/indexer.h"
+#include "search/search_encoding.h"
+#include "storage/redis_metadata.h"
+#include "storage/storage.h"
+
+
+namespace redis {
+
+struct VectorItem {
+  std::string key;
+  // TODO: use template based on VectorType
+  std::vector<double> vector;
+  HnswVectorFieldMetadata* metadata;
+
+  VectorItem(std::string_view key, std::string_view vector_str, HnswVectorFieldMetadata* metadata) : key(key), metadata(metadata) { 
+    Decode(vector_str); 
+  }
+
+  // TODO: move it to util
+  void Decode(std::string_view vector_str) {
+    std::string trimmed = std::string(vector_str);
+    trimmed.erase(0, 1);                   // remove the first '['
+    trimmed.erase(trimmed.size() - 1, 1);  // remove the last ']'
+
+    std::istringstream iss(trimmed);
+    std::string num;
+
+    vector.clear();
+
+    while (std::getline(iss, num, ',')) {
+      try {
+        double value = std::stod(num);
+        vector.push_back(value);
+      } catch (const std::invalid_argument& ia) {
+        throw std::runtime_error("Invalid number in vector string: " + num);
+      }
+    }
+  }
```

Comment:
It appears that our main issue here is not implementing a value system in KQIR, but relying on strings to represent values of all field types. 

Initially, I avoided introducing it for quick implementation at the outset. However, it now seems essential.

In HASH type vector fields, data is encoded in binary format; whereas in JSON, it is stored as an array of numbers. (https://redis.io/docs/latest/develop/interact/search-and-query/advanced-concepts/vectors/#storing-vectors-in-hashes)

Reply:
I've been exploring the possibility of storing distances in metadata to enhance computation efficiency, but given that these distances are always calculated between random vectors, I haven't yet discovered an effective work-around to reduce the number of compute. 

Thanks for pointing out the different data formats! If that's the case, I'm wondering if adopting the binary format to encode the vector for both JSON and HASH would be a good idea? This can save space and improve performance compared to storing them as raw string for now, but has some overhead when the vector is passed in as an array in JSON. Let me know if you prefer this solution or not :))

Reply:
Now `kqir::Value` is available, and `IndexUpdater::UpdateXXXIndex` can take `kqir::Value` as input.

For encoding part, we cannot change the existing encoding of JSON and HASH data types. But you can store the vector in binary format in the HNSW index encoding.

https://github.com/apache/kvrocks/blob/unstable/src/search/value.h

Reply:
- Add a new template argument to the base class (std::variant) of kqir::Value (e.g. NumericArray)
- Modify FieldValueRetriever::Retrieve to handle the case while the input field type is HnswVectorFieldMetadata

Done in #2371.